stat/stat.go

```diff
@@ -0,0 +1,59 @@
+package stat
+
+import (
+	"sync"
+	"time"
+)
+
+type Stat struct {
+	// requestTimes is a map of hour -> method -> count
+	requestTimes map[int]map[string]int
+	// initTimes is a map of hour -> time
+	initTimes  map[int]time.Time
+	requestsMu *sync.Mutex
+}
+
+// day is the number of nanoseconds in a day
+const day = 24 * time.Hour
+
+func NewStat() Stat {
+	requestTimes := make(map[int]map[string]int)
+	initTimes := make(map[int]time.Time)
+	return Stat{
+		requestTimes: requestTimes,
+		initTimes:    initTimes,
+		requestsMu:   &sync.Mutex{},
+	}
+}
+
+func (stat *Stat) Get() map[string]int {
+	t := time.Now()
+	numRequests := make(map[string]int)
+	stat.requestsMu.Lock()
+	for hour, methodTimestamps := range stat.requestTimes {
+		// only count the hour if it was within the past day
+		if t.Sub(stat.initTimes[hour]) > day {
+			continue
+		}
+
+		for method, count := range methodTimestamps {
+			numRequests[method] += count
+		}
+	}
+	stat.requestsMu.Unlock()
+	return numRequests
+}
+
+func (stat *Stat) Insert(method string) {
+	t := time.Now()
+	stat.requestsMu.Lock()
+
+	// initialise the secondary map if nil
+	if stat.requestTimes[t.Hour()] == nil || t.Sub(stat.initTimes[t.Hour()]) > day {
```

Comment:
I think the logic isn't quite right here. If I'm understanding correctly the goal is to count the number of occurrences of `method` for the past day grouping by the hour, but the following is an example where I think it will be not quite right:

- The first time method `x` is seen on day 1 is at 0030
- In this case `stat.initTimes[t.Hour()]` is 0030 on day 1
- On day 2 the method `x` is seen at 0015
- This means that `t.Sub(stat.initTimes[t.Hour()])` will actually be less than `day` and won't trigger a reinitialisation, which I think it should

In this case I would think that the correct logic is to compare the time difference to `time.Hour`.

Reply:
`t` is now truncated to the hour so the logic should be correct now.